FWCore/FWCore/DataHandle.h

```diff
@@ -61,9 +80,40 @@ DataHandle<T>::DataHandle() {}
  */
 template<typename T>
 const T* DataHandle<T>::get() {
-  return DataObjectHandle<DataWrapper<T> >::get()->getData();
+  DataObject* dataObjectp = nullptr;
+  // once we migrate to Gaudi v27r1+ replace with this:
+  // DataObjectHandle<DataWrapper<T>>::m_EDS->retrieveObject(DataObjectHandle<DataWrapper<T> >::dataProductName(), 
+  //    dataObjectp);
+  auto sc = DataHandle<T>::m_eds->retrieveObject(DataObjectHandle<DataWrapper<T> >::dataProductName(), dataObjectp);
+  if (sc.isFailure() && !DataObjectHandle<DataWrapper<T>>::m_descriptor->alternativeAddresses().empty()) {
+    for(uint i = 0; i < DataObjectHandle<DataWrapper<T>>::m_descriptor->alternativeAddresses().size() && sc.isFailure(); ++i){
+      sc = m_eds->retrieveObject(DataObjectHandle<DataWrapper<T>>::m_descriptor->alternativeAddresses()[i], dataObjectp);
+    }
+  }
+
+  if (LIKELY(sc.isSuccess())) {
+    if (UNLIKELY(!m_isGoodType)) {
+      // done once for write objects, each time if created by reader
+      m_isGoodType = nullptr != dynamic_cast<DataWrapper<T>*> (dataObjectp);
+    }
+    if (LIKELY(m_isGoodType)) {
+      DataObjectHandle<DataWrapper<T> >::setRead();
+      std::cout << "returning static cast" << std::endl;
+      return static_cast<DataWrapper<T>*>(dataObjectp)->getData();
+    } else {
+      // The reader does not know the specific type of the collection. So we need a reinterpret_cast if the handle was
+      // created by the reader.
+      DataWrapper<podio::CollectionBase>* tmp = static_cast<DataWrapper<podio::CollectionBase>*>(dataObjectp);
+      DataObjectHandle<DataWrapper<T> >::setRead();
+      return reinterpret_cast<const T*>(tmp->collectionBase());
```

Comment:
I added `dynamic_casts` (done once) further up to check if it is safe to do the `static_cast` and `reinterpret_cast` here.